### ISSUE TITLE
Introduce the Generic.xaml concept

### ISSUE THREAD

**Issue by dotMorten** [dotnet/maui#9612]
### Description

As a control developer I often write controls that have templates. Templates are great because it allows the developer to overwrite the styles easily. However templating custom controls in MAUI is surprisingly hard compared to WPF, UWP and WinUI.
In those frameworks all you need to do is include a `Themes\Generic.xaml` file with the style, and in each control set the `DefaultStyleKey` property. At this point the Generic.xaml styles are automatically loaded and applied to these controls, unless the developer chooses to overwrite these styles. It makes for a very clean story for creating custom control libraries.

### Public API Changes

```csharp
public class MyControl : TemplatedView
{
   public MyControl()
   {
       DefaultStyleKey = typeof(MyControl);
   }
}

````

### Intended Use-Case

As a control developer, I want to build class libraries with custom controls in it, and these controls inherit from TemplatedView. I want to use the XAML designed to define the styles, and I want to make it easy for end developers to also overwrite these styles.
Coming from the other XAML platforms, we're used to do this with the `Generic.xaml` file in the `Themes` folder. A similar concept is missing from .NET MAUI.
There we're forced to either: 
 - manually load it in the constructor from a string (slow) [example](https://github.com/Esri/arcgis-toolkit-dotnet/blob/754d62a66fff9245cbd825f656cde5a0d74a65ff/src/Toolkit/Toolkit.Maui/ScaleLine/ScaleLine.cs#L37-L86)
 - ask the end user to add a resource dictionary to their application (error-prone / cumbersome)
 - On-the-fly add insert our own resource dictionary into the application resources, which is error prone (could be added in the wrong order and override user-defined styles, and it feels wrong for a control to be modifying application resources).  [example - literally called a hack](https://dotnetdevaddict.co.za/2020/08/16/templated-controls-in-xamarin-forms/#hack)

**Comment by PureWeen** [comment]
related

https://github.com/dotnet/maui/issues/6350
https://github.com/dotnet/maui/issues/4132

**Comment by vanka78bg** [comment]
Please, make this high priority. Currently, component authors have to resort to ugly hacks to provide some default styles and theming for the controls.

**Comment by LanceMcCarthy** [comment]
Bump 👍🏼  Are there any branches this is being worked on that I can follow/experiment with?

**Comment by Alex-Dobrynin** [comment]
+1 UP

**Comment by mattleibow** [comment]
I don't think we have started any work on this. However, one issue to deal with is the "scanning" part. 

1. If this happens at runtime, we have to load all assemblies which is not what we want to do I am thinking? 
2. If we do it at compile time, then we have to scan all assemblies and add references - but then we now reference everything and the linker will be leaving more things behind.

However, I think option 2 is a bit better since it is faster for startup and the code can all be generated.

@StephaneDelcroix are there any features that we have today that will do things for us or make things easier? Is there a way to hook into the xaml build?

**Comment by panayot-cankov** [comment]
@mattleibow There are more details in how Generic.xaml worked in WPF

"[DefaultStyleKey](https://learn.microsoft.com/en-us/dotnet/api/system.windows.frameworkelement.defaultstylekey?view=windowsdesktop-8.0)" would be used to look up a "default style" in WPF. Its values would apply with lower precedence than an "implicit style".

"[ResourceKey](https://learn.microsoft.com/en-us/dotnet/api/system.windows.resourcekey.assembly?view=windowsdesktop-8.0)" could be created by a control instance for its "DefaultStyleKey", and the "ResourceKey" has an Assembly property. When WPF would lookup Generic.xaml, it will do so in the Assembly specified by that property. That determines which assemblies have to be scanned for "Generic.xaml".

The above two features in combination, allowed for Control developers to ship "theme assemblies":
 - Application developers could select a theme per vendor based on user settings
 - The theme could change at runtime, changing the "default style"
 - The application developer's implicit styles would still apply on top of default (theme) styles

Linker is a bummer, for .NET MAUI I'd be awesome if we could still ship multiple assemblies with themes containing Generic.xaml but have them stripped from unused styles after the linking phase.

**Comment by AathifMahir** [comment]
Can we expect this for Net 9 version of Maui?
